internal/controller/tlsconfigutil/tls_config_util_test.go

```diff
@@ -0,0 +1,661 @@
+				&corev1.ConfigMap{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      "awesome-cm",
+						Namespace: "awesome-namespace",
+					},
+					Data: map[string]string{
+						"ca-bundle": string(bundle),
+					},
+				},
+			},
+			expectedCondition: &metav1.Condition{
+				Type:    typeTLSConfigurationValid,
+				Status:  metav1.ConditionFalse,
+				Reason:  ReasonInvalidTLSConfig,
+				Message: "tls.certificateAuthorityDataSource is invalid: unsupported CA bundle source kind: SomethingElse",
+			},
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			t.Parallel()
+			var secretsInformer corev1informers.SecretInformer
+			var configMapInformer corev1informers.ConfigMapInformer
+
+			if len(tt.k8sObjects) > 0 {
+				fakeClient := fake.NewSimpleClientset(tt.k8sObjects...)
+				sharedInformers := informers.NewSharedInformerFactory(fakeClient, 0)
+				configMapInformer = sharedInformers.Core().V1().ConfigMaps()
+				secretsInformer = sharedInformers.Core().V1().Secrets()
+
+				// calling the .Informer function registers this informer in the sharedinformer.
+				// doing this will ensure that this informer will be sync'd when Start is called next.
+				configMapInformer.Informer()
+				secretsInformer.Informer()
+
+				ctx, cancel := context.WithCancel(context.Background())
+				defer cancel()
+				sharedInformers.Start(ctx.Done())
+				sharedInformers.WaitForCacheSync(ctx.Done())
+			}
+			actualCondition, _, _, _ := ValidateTLSConfig(tt.tlsSpec, "tls", tt.namespace, secretsInformer, configMapInformer)
+			require.Equal(t, tt.expectedCondition, actualCondition)
+		})
+	}
+}
+
+func TestReadCABundleFromK8sSecret(t *testing.T) {
+	tests := []struct {
+		name            string
+		secretNamespace string
+		secretName      string
+		secretKey       string
+		k8sObjects      []runtime.Object
+		expectedData    string
+		expectError     bool
+	}{
+		{
+			name:            "should return error reading a non-existent secret",
+			secretNamespace: "awesome-namespace",
+			secretName:      "does-not-exist",
+			secretKey:       "does-not-matter",
+			k8sObjects: []runtime.Object{
+				&corev1.Secret{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      "awesome-secret",
+						Namespace: "awesome-namespace",
+					},
+					Data: map[string][]byte{
+						"awesome": []byte("pinniped-is-awesome"),
+					},
+				},
+			},
+			expectedData: "",
+			expectError:  true,
+		},
+		{
+			name:            "should return error reading a non-existing key in an existing secret",
+			secretNamespace: "awesome-namespace",
+			secretName:      "awesome-secret",
+			secretKey:       "something-else",
+			k8sObjects: []runtime.Object{
+				&corev1.Secret{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      "awesome-secret",
+						Namespace: "awesome-namespace",
+					},
+					Data: map[string][]byte{
+						"awesome": []byte("pinniped-is-awesome"),
+					},
+				},
+			},
+			expectedData: "",
+			expectError:  true,
```

Comment:
Could we check errors by matching a string? I think this test is passing for the wrong reason.

Reply:
Done!